sql/src/main/java/io/crate/metadata/doc/CrateMetaDataUpgradeService.java

```diff
@@ -0,0 +1,146 @@
+            logger.error("template={} could not be upgraded", indexTemplateMetaData.getName());
+            throw new RuntimeException(e);
+        }
+    }
+
+    private IndexMetaData upgradeIndexMapping(IndexMetaData indexMetaData) throws IOException {
+        MappingMetaData mappingMetaData = indexMetaData.mapping(Constants.DEFAULT_MAPPING_TYPE);
+        MappingMetaData newMappingMetaData = saveRoutingHashFunctionToMapping(
+            mappingMetaData,
+            indexMetaData.getSettings().get(IndexMetaData.SETTING_LEGACY_ROUTING_HASH_FUNCTION));
+        if (mappingMetaData != newMappingMetaData) {
+            logger.info("upgraded mapping of index={}", indexMetaData.getIndex());
+            return IndexMetaData.builder(indexMetaData)
+                .removeMapping(Constants.DEFAULT_MAPPING_TYPE)
+                .putMapping(newMappingMetaData)
+                .build();
+        }
+        return indexMetaData;
+    }
+
+    private IndexTemplateMetaData upgradeTemplateMapping(IndexTemplateMetaData indexTemplateMetaData) throws IOException {
+        // we only want to upgrade partition table related templates
+        if (PartitionName.isPartition(indexTemplateMetaData.getTemplate()) == false) {
+            return indexTemplateMetaData;
+        }
+        MappingMetaData mappingMetaData = new MappingMetaData(indexTemplateMetaData.getMappings().get(Constants.DEFAULT_MAPPING_TYPE));
+        MappingMetaData newMappingMetaData = saveRoutingHashFunctionToMapping(
+            mappingMetaData,
+            indexTemplateMetaData.getSettings().get(IndexMetaData.SETTING_LEGACY_ROUTING_HASH_FUNCTION));
+        if (mappingMetaData != newMappingMetaData) {
+            logger.info("upgraded mapping of template={}", indexTemplateMetaData.getName());
+            return new IndexTemplateMetaData.Builder(indexTemplateMetaData)
+                .removeMapping(Constants.DEFAULT_MAPPING_TYPE)
+                .putMapping(Constants.DEFAULT_MAPPING_TYPE, newMappingMetaData.source())
+                .build();
+        }
+        return indexTemplateMetaData;
+    }
+
+    private MappingMetaData saveRoutingHashFunctionToMapping(@Nullable MappingMetaData mappingMetaData,
+                                                             @Nullable String routingHashFunction) throws IOException {
+        Map<String, Object> mappingMap = null;
+        if (mappingMetaData != null) {
+            mappingMap = mappingMetaData.sourceAsMap();
+        }
+        String hashFunction = DocIndexMetaData.getRoutingHashFunction(mappingMap);
+        if (hashFunction == null) {
+            if (routingHashFunction == null) {
+                routingHashFunction = DocIndexMetaData.DEFAULT_ROUTING_HASH_FUNCTION;
+            }
+            // create new map, existing one can be immutable
+            Map<String, Object> newMappingMap = mappingMap == null
+                ? new HashMap<>(1)
+                : MapBuilder.newMapBuilder(mappingMap).map();
```

Comment:
Minor thing: could also use `new HashMap<>(mappingMap)` to avoid the intermediate MapBuilder instance